From 6ac95c963d32182791c806fcd23e649ce0393937 Mon Sep 17 00:00:00 2001
From: d34d <clark@cyngn.com>
Date: Tue, 25 Aug 2015 19:13:46 -0700
Subject: [PATCH 172/296] Show icon of package associated with Toast

For all those times you have some random app or background service
that posts a Toast and you have no idea who's posting it.  This adds
an icon badge to the top left corner of the Toast to show the app's
icon the Toast belongs to.
---
 core/java/android/widget/Toast.java            | 14 ++++++++++++++
 core/res/res/layout/transient_notification.xml | 23 ++++++++++++++++-------
 2 files changed, 30 insertions(+), 7 deletions(-)

diff --git a/core/java/android/widget/Toast.java b/core/java/android/widget/Toast.java
index 5addaa9b420..a1bcb47e456 100644
--- a/core/java/android/widget/Toast.java
+++ b/core/java/android/widget/Toast.java
@@ -23,9 +23,11 @@ import android.annotation.StringRes;
 import android.app.INotificationManager;
 import android.app.ITransientNotification;
 import android.content.Context;
+import android.content.pm.PackageManager;
 import android.content.res.Configuration;
 import android.content.res.Resources;
 import android.graphics.PixelFormat;
+import android.graphics.drawable.Drawable;
 import android.os.Handler;
 import android.os.IBinder;
 import android.os.Looper;
@@ -463,6 +465,18 @@ public class Toast {
                 if (context == null) {
                     context = mView.getContext();
                 }
+
+                ImageView appIcon = (ImageView) mView.findViewById(android.R.id.icon);
+                if (appIcon != null) {
+                    PackageManager pm = context.getPackageManager();
+                    Drawable icon = null;
+                    try {
+                        icon = pm.getApplicationIcon(packageName);
+                    } catch (PackageManager.NameNotFoundException e) {
+                        // nothing to do
+                    }
+                    appIcon.setImageDrawable(icon);
+                }
                 mWM = (WindowManager)context.getSystemService(Context.WINDOW_SERVICE);
                 // We can resolve the Gravity here by using the Locale for getting
                 // the layout direction
diff --git a/core/res/res/layout/transient_notification.xml b/core/res/res/layout/transient_notification.xml
index db586ec37cf..49b7e9d7042 100644
--- a/core/res/res/layout/transient_notification.xml
+++ b/core/res/res/layout/transient_notification.xml
@@ -18,24 +18,33 @@
 */
 -->
 
-<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="match_parent"
     android:layout_height="match_parent"
     android:orientation="vertical"
-    android:background="?android:attr/toastFrameBackground">
+    android:clipChildren="false">
 
     <TextView
         android:id="@android:id/message"
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
-        android:layout_weight="1"
-        android:layout_marginHorizontal="24dp"
-        android:layout_marginVertical="15dp"
+        android:paddingHorizontal="24dp"
+        android:paddingVertical="15dp"
         android:layout_gravity="center_horizontal"
+        android:layout_marginTop="-16dp"
+        android:layout_marginStart="-16dp"
+        android:layout_toEndOf="@android:id/icon"
+        android:layout_below="@android:id/icon"
         android:textAppearance="@style/TextAppearance.Toast"
         android:textColor="@color/primary_text_default_material_light"
+        android:background="?android:attr/toastFrameBackground"
         />
 
-</LinearLayout>
-
+    <ImageView
+        android:id="@android:id/icon"
+        android:layout_width="24dp"
+        android:layout_height="24dp"
+        android:layout_alignParentTop="true"
+        android:layout_alignParentStart="true"/>
 
+</RelativeLayout>
-- 
2.11.0

